source/extensions/network/dns_resolver/cares/dns_impl.cc

```diff
@@ -89,10 +116,20 @@ void DnsResolverImpl::initializeChannel(ares_options* options, int optmask) {
     static_cast<DnsResolverImpl*>(arg)->onAresSocketStateChange(fd, read, write);
   };
   options->sock_state_cb_data = this;
-  ares_init_options(&channel_, options, optmask | ARES_OPT_SOCK_STATE_CB);
+  optmask |= ARES_OPT_SOCK_STATE_CB;
+
+  ENVOY_LOG_MISC(error, "resolvers {}, use dns fallback {}", resolvers_vector_.size(),
+                 use_resolvers_as_fallback_);
+  if (!resolvers_vector_.empty() && use_resolvers_as_fallback_) {
+    ENVOY_LOG_MISC(error, "using resolvers as fallback {}", resolvers_vector_.size());
+    options->nservers = resolvers_vector_.size();
+    options->servers = &resolvers_vector_[0];
+    optmask |= ARES_OPT_SERVERS;
+  }
+  ares_init_options(&channel_, options, optmask);
 
   // Ensure that the channel points to custom resolvers, if they exist.
-  if (resolvers_csv_.has_value()) {
+  if (resolvers_csv_.has_value() && !use_resolvers_as_fallback_) {
```

Comment:
Should this case be blocked at config load? It seems like a config error that might otherwise be confusing?